pydantic_ai_slim/pydantic_ai/models/google.py

```diff
@@ -368,20 +369,82 @@ def _get_tools(
                     raise UserError(
                         f'`{tool.__class__.__name__}` is not supported by `GoogleModel`. If it should be, please file an issue.'
                     )
-        return tools or None, image_config
+        return tools, image_config
 
     def _get_tool_config(
-        self, model_request_parameters: ModelRequestParameters, tools: list[ToolDict] | None
-    ) -> ToolConfigDict | None:
-        if not model_request_parameters.allow_text_output and tools:
-            names: list[str] = []
-            for tool in tools:
-                for function_declaration in tool.get('function_declarations') or []:
-                    if name := function_declaration.get('name'):  # pragma: no branch
-                        names.append(name)
-            return _tool_config(names)
+        self,
+        model_request_parameters: ModelRequestParameters,
+        model_settings: GoogleModelSettings,
+    ) -> tuple[list[ToolDict] | None, ToolConfigDict | None, ImageConfigDict | None]:
+        """Determine which tools to send and the API tool config.
+
+        Returns:
+            A tuple of (filtered_tools, tool_config, image_config).
+        """
+        function_tools = model_request_parameters.function_tools
+        output_tools = model_request_parameters.output_tools
+        builtin_tools, image_config = self._get_builtin_tools(model_request_parameters)
+
+        resolved = _resolve_tool_choice(model_settings, model_request_parameters)
+
+        if resolved is None:
+            tool_defs_to_send = [*function_tools, *output_tools]
         else:
-            return None
+            tool_defs_to_send = resolved.filter_tools(function_tools, output_tools)
```

Comment:
Since Google supports a list of allowed function names, I don't think we should filter at all